futures-02.md

```diff
@@ -0,0 +1,396 @@
+These will be addressed in later sections.
+
+[current]: https://docs.rs/futures/0.1.16/futures/task/fn.current.html
+[notify]: https://docs.rs/futures/0.1.16/futures/task/struct.Task.html#method.notify
+
+## Crate Separation
+
+The `futures` crate will be split into several different libraries.
+
+### `futures-core`
+
+`futures-core` will include `Future`, `Stream`, `IntoFuture`
+traits and their dependencies. This is a central crate that includes only
+essential items, designed to minimize future breakage. Combinator methods
+will be removed from the `Future` and `Stream` and moved to the
+`FutureExt` and `StreamExt` traits in the `futures-util` crate. `poll` 
+will be the only remaining method on the `Future` and `Stream` traits.
+This separation allows combinator functions to evolve over time without
+breaking changes to the public API surface of async libraries.
+
+### `futures-io`
+
+`futures-io` includes the updated `AsyncRead` and `AsyncWrite` traits.
+These traits were previously included in the `tokio-io` crate.
+However, `AsyncRead` and `AsyncWrite` don't have any dependencies on the `tokio`
+crate or the rest of the `tokio` stack, so they will be moved into the `futures`
+group of libraries. `futures-io` depends only on the `futures-core` crate
+and the `bytes` crate (for the [`Buf`] and [`BufMut`] traits).
+
+[`Buf`]: https://carllerche.github.io/bytes/bytes/trait.Buf.html
+[`BufMut]: https://carllerche.github.io/bytes/bytes/trait.BufMut.html
+
+### `futures-sink`
+
+`futures-sink` includes the `Sink` trait, which allows values to be sent
+asynchronously. The `Sink` trait is excluded from `futures-core` because its
+design is much less stable than `Future` and `Stream`. There are several
+potential breaking changes to the `Sink` trait under consideration.
+
+This crate will depend only on `futures-core`.
+
+### `futures-channel`
+
+`futures-channel` includes the `mpsc` and `oneshot` messaging primitives.
+This crate is a minimal addition on `futures-core` designed for low churn
+so that channels may appear in the public APIs of libraries.
+
+### `futures-util`
+
+`futures-util` contains a number of convenient combinator functions for
+`Future`, `Stream`, and `Sink`, including the `FutureExt`, `StreamExt`,
+and `SinkExt` extension traits.
+
+This crate depends on `futures-core`, `futures-io`, and `future-sink`.
+
+### `futures-executor`
+
+`futures-executor` contains the `Executor` trait which allows spawning
+asynchronously tasks onto a task executor. Like `Sink`, this is a less
+mature API which hasn't seen as much practical use. There are several
+possible breaking changes to this API coming in the near future.
+
+This crate depends on `futures-core`.
+
+## Small, Breaking API Improvements and Clarifications
+
+- `Task` will be renamed to `Waker`. This name more clearly reflects how
+  the type is used. It is a handle used to awaken tasks-- it is not a `Task`
+  itself.
+- `Async` variant `NotReady` will be renamed to `Pending`.
```

Comment:
With the new `Try` trait, would it make sense to explore instead of returning `Result<Async<T>, E>`, making `Async` an enum of 3 variants?

Reply:
How would an enum with 3 variants work with `Try`? Can you elaborate?